File: src/main/java/org/lanternpowered/server/entity/LanternTransform.java
package org.lanternpowered.server.entity;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import javax.annotation.Nullable;

import org.spongepowered.api.entity.Transform;
import org.spongepowered.api.world.Location;
import org.spongepowered.api.world.extent.Extent;

import com.flowpowered.math.imaginary.Quaterniond;
import com.flowpowered.math.matrix.Matrix4d;
import com.flowpowered.math.vector.Vector3d;

public class LanternTransform <E extends Extent> implements Transform<E> {

    @Nullable private E extent;

    private Vector3d position;
    private Vector3d rotation;
    private Vector3d scale;

    @Nullable private Location<E> location = null;
    @Nullable private Quaterniond rotationQuaternion = null;

    public LanternTransform() {
        this(null, Vector3d.ZERO, Vector3d.ZERO, Vector3d.ONE);
    }

    public LanternTransform(@Nullable E extent, Vector3d position) {
        this(extent, position, Vector3d.ZERO, Vector3d.ONE);
    }

    public LanternTransform(@Nullable E extent, Vector3d position, Vector3d rotation, Vector3d scale) {
        this.position = checkNotNull(position, "position");
        this.rotation = checkNotNull(rotation, "rotation");
        this.scale = checkNotNull(scale, "scale");
        this.extent = extent;
    }

    @Override
    public Location<E> getLocation() {
        checkState(extent != null, "Transform has no extent");
        if (this.location == null) {
            this.location = new Location<E>(this.extent, this.position);
        }
        return this.location;
    }

    @Override
    public Transform<E> setLocation(Location<E> location) {
        checkNotNull(location, "location");
        this.setExtent(location.getExtent());
        this.setPosition(location.getPosition());
        return this;
    }

    @Override
    public E getExtent() {
        checkState(extent != null, "Transform has no extent");
        return this.extent;
    }

    @Override
    public Transform<E> setExtent(E extent) {
        checkNotNull(extent, "extent");
        this.extent = extent;
        this.location = null;
        return this;
    }

    @Override
    public Vector3d getPosition() {
        return this.position;
    }

    @Override
    public Transform<E> setPosition(Vector3d position) {
        checkNotNull(position, "position");
        this.position = position;
        this.location = null;
        return this;
    }

    @Override
    public Vector3d getRotation() {
        return this.rotation;
    }

    @Override
    public Transform<E> setRotation(Vector3d rotation) {
        checkNotNull(rotation, "rotation");
        this.rotation = rotation;
        this.rotationQuaternion = null;
        return this;
    }

    @Override
    public Quaterniond getRotationAsQuaternion() {
        if (this.rotationQuaternion == null) {
            this.rotationQuaternion = Quaterniond.fromAxesAnglesDeg(
                    this.rotation.getX(), -this.rotation.getY(), this.rotation.getZ());
        }
        return this.rotationQuaternion;
    }

    @Override
    public Transform<E> setRotation(Quaterniond rotation) {
        checkNotNull(rotation, "rotation");
        final Vector3d axesAngles = rotation.getAxesAnglesDeg();
        this.rotation = new Vector3d(axesAngles.getX(), -axesAngles.getY(), axesAngles.getZ());
        this.rotationQuaternion = rotation;
        return this;
    }

    @Override
    public double getPitch() {
        return this.rotation.getX();
    }

    @Override
    public double getYaw() {
        return this.rotation.getY();
    }

    @Override
    public double getRoll() {
        return this.rotation.getZ();
    }

    @Override
    public Vector3d getScale() {
        return this.scale;
    }

    @Override
    public Transform<E> setScale(Vector3d scale) {
        checkNotNull(scale, "scale");
        this.scale = scale;
        return this;
    }

    @Override
    public Transform<E> add(Transform<E> other) {
        checkNotNull(other, "other");
        this.addTranslation(other.getPosition());
        this.addRotation(other.getRotationAsQuaternion());
        this.addScale(other.getScale());
        return this;
    }

    @Override
    public Transform<E> addTranslation(Vector3d translation) {
        checkNotNull(translation, "translation");
        this.setPosition(this.getPosition().add(translation));
        return this;
    }

    @Override
    public Transform<E> addRotation(Vector3d rotation) {
        checkNotNull(rotation, "rotation");
        return this.addRotation(Quaterniond.fromAxesAnglesDeg(
                rotation.getX(), -rotation.getY(), rotation.getZ()));
    }

    @Override
    public Transform<E> addRotation(Quaterniond rotation) {
        checkNotNull(rotation, "rotation");
        this.setRotation(rotation.mul(this.getRotationAsQuaternion()));
        return this;
    }

    @Override
    public Transform<E> addScale(Vector3d scale) {
        checkNotNull(scale, "scale");
        this.setScale(this.getScale().mul(scale));
        return this;
    }

    @Override
    public Matrix4d toMatrix() {
        return Matrix4d.createScaling(this.getScale().toVector4(1)).rotate(this.getRotationAsQuaternion())
                .translate(this.getPosition());
    }

    @Override
    public boolean isValid() {
        return this.extent != null && this.extent.isLoaded();
    }

    @Override
    public String toString() {
        return "Transform{location=" + this.getLocation() + ", rotation=" + this.getRotation() +
                ", scale=" + this.getScale() + '}';
    }
}

File: src/main/kotlin/org/lanternpowered/server/util/LanternTransform.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.util

import org.lanternpowered.api.util.ToStringHelper
import org.spongepowered.api.util.Transform
import org.spongepowered.math.imaginary.Quaterniond
import org.spongepowered.math.matrix.Matrix4d
import org.spongepowered.math.vector.Vector3d

class LanternTransform @JvmOverloads constructor(
        private val position: Vector3d,
        private val rotation: Vector3d = Vector3d.ZERO,
        private val scale: Vector3d = Vector3d.ONE
) : Transform {

    private val rotationQuaternion by lazy { fromAxesAngles(this.rotation) }

    override fun getPosition() = this.position
    override fun getRotation() = this.rotation
    override fun getRotationAsQuaternion() = this.rotationQuaternion
    override fun getScale() = this.scale

    override fun getPitch() = this.rotation.x
    override fun getYaw() = this.rotation.y
    override fun getRoll() = this.rotation.z

    override fun withPosition(position: Vector3d) = LanternTransform(position, this.rotation, this.scale)
    override fun withRotation(rotation: Quaterniond) = withRotation(toAxesAngles(rotation))
    override fun withRotation(rotation: Vector3d) = LanternTransform(this.position, rotation, this.scale)
    override fun withScale(scale: Vector3d) = LanternTransform(this.position, this.rotation, scale)

    override fun add(other: Transform): Transform {
        val position = this.position.add(other.position)
        val rotation = toAxesAngles(other.rotationAsQuaternion.mul(this.rotationAsQuaternion))
        val scale = this.scale.mul(other.scale)
        return LanternTransform(position, rotation, scale)
    }

    override fun translate(translation: Vector3d) = withPosition(this.position.add(translation))
    override fun rotate(rotation: Vector3d) = rotate(fromAxesAngles(rotation))
    override fun rotate(rotation: Quaterniond) = withRotation(toAxesAngles(rotation.mul(this.rotationAsQuaternion)))
    override fun scale(scale: Vector3d) = withScale(this.scale.mul(scale))

    override fun toMatrix(): Matrix4d = Matrix4d.createScaling(this.scale.toVector4(1f))
            .rotate(this.rotationAsQuaternion).translate(this.position)

    override fun hashCode(): Int {
        var result = this.position.hashCode()
        result = 31 * result + this.rotation.hashCode()
        result = 31 * result + this.scale.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is Transform) {
            return false
        }
        return other.position == this.position && other.rotation == this.rotation && other.scale == this.scale
    }

    private fun toAxesAngles(quaternion: Quaterniond): Vector3d {
        val axesAngles = quaternion.axesAnglesDeg
        return Vector3d(axesAngles.x, -axesAngles.y, axesAngles.z)
    }

    private fun fromAxesAngles(angles: Vector3d): Quaterniond = Quaterniond.fromAxesAnglesDeg(angles.x, -angles.y, angles.z)

    override fun toString() = ToStringHelper("Transform")
            .add("position", this.position)
            .add("rotation", this.rotation)
            .add("scale", this.scale)
            .toString()

    object Factory : Transform.Factory {

        override fun create(position: Vector3d, rotation: Vector3d, scale: Vector3d)
                = LanternTransform(position, rotation, scale)
    }
}
